src/core/util/byte_stream.cc

```diff
@@ -60,55 +61,77 @@ std::uint8_t* InputByteStream::ReadBytes(
   return ptr;
 }
 
+std::uint64_t InputByteStream::ReadUInt64() {
+  return bufbe64toh(ReadBytes(sizeof(std::uint64_t)));
+}
+
 std::uint32_t InputByteStream::ReadUInt32() {
-  return bufbe32toh(ReadBytes(4));
+  return bufbe32toh(ReadBytes(sizeof(std::uint32_t)));
 }
 
 std::uint16_t InputByteStream::ReadUInt16() {
-  return bufbe16toh(ReadBytes(2));
+  return bufbe16toh(ReadBytes(sizeof(std::uint16_t)));
 }
 
 std::uint8_t InputByteStream::ReadUInt8() {
-  return *ReadBytes(1);
+  return *ReadBytes(sizeof(std::uint8_t));
 }
 
 OutputByteStream::OutputByteStream(
   std::uint8_t* data,
   std::size_t len)
-  : m_Data(data), m_Length(len) { }
+  : m_Data(data),
+    m_Length(len),
+    m_Counter(0),
+    m_Size(len) {}
 
 void OutputByteStream::ProduceData(std::size_t amount) {
   if (amount > m_Length)
-    throw std::length_error("SSUPacketParser: too many bytes to produce.");
+    throw std::length_error("OutputByteStream: too many bytes to produce.");
   m_Data += amount;
   m_Length -= amount;
+  m_Counter += amount;
 }
 
 void OutputByteStream::WriteData(const std::uint8_t* data, std::size_t len) {
   std::uint8_t* ptr = m_Data; 
   ProduceData(len);
-  memcpy(ptr, data, len);
+  std::memcpy(ptr, data, len);
 }
 
 void OutputByteStream::WriteUInt8(std::uint8_t data) {
-  WriteData(&data, 1);
+  WriteData(&data, sizeof(std::uint8_t));
 }
 
 void OutputByteStream::WriteUInt16(std::uint16_t data) {
-  std::uint8_t buf[2] = {};
+  std::uint8_t buf[sizeof(std::uint16_t)] = {};
   htobe16buf(buf, data);
   WriteData(buf, sizeof(buf));
 }
 
 void OutputByteStream::WriteUInt32(std::uint32_t data) {
-  std::uint8_t buf[4] = {};
+  std::uint8_t buf[sizeof(std::uint32_t)] = {};
   htobe32buf(buf, data);
   WriteData(buf, sizeof(buf));
 }
 
+void OutputByteStream::WriteUInt64(std::uint64_t data) {
+  std::uint8_t buf[sizeof(std::uint64_t)] = {};
+  htobe64buf(buf, data);
+  WriteData(buf, sizeof(buf));
+}
+
 std::uint8_t* OutputByteStream::GetPosition() const {
   return m_Data;
 }
 
+std::uint8_t* OutputByteStream::GetData() const {
```

Comment:
@EinMByte wrote these bytestream classes, so maybe he has his reasons. As for `TunnelHopConfig`, I almost *always* place getters/setters in header but I specifically chose not to for `TunnelHopConfig` purely out of aesthetics and maintainability (tunnel logic is tricky at times and I needed a clean slate to work on). I'm flexible in either case though because #511 is now open and there is a good trade-off in favor of optimization in this area.